src/engine/Engine.js

```diff
@@ -0,0 +1,381 @@
+    if (!this._runtimeErrors[cellId]) this._runtimeErrors[cellId] = {}
+    if (isArray(error)) {
+      error = error.map(_normalizeError)
+    } else {
+      error = _normalizeError(error)
+    }
+    this._runtimeErrors[cellId][key] = error
+
+    function _normalizeError(error) {
+      const line = error.hasOwnProperty('line') ? error.line : -1
+      const column = error.hasOwnProperty('column') ? error.column : -1
+      const message = error.message || error.toString()
+      return {line, column, message}
+    }
+  }
+
+  // called by custom components, such as VegaLiteComponent
+  clearRuntimeError(cellId, key) {
+    const errors = this._runtimeErrors[cellId]
+    if (errors) {
+      delete errors[key]
+    }
+  }
+
+  hasSyntaxError(cellId) {
+    const expr = this.getExpression(cellId)
+    return expr && Boolean(expr.syntaxError)
+  }
+
+  getSyntaxError(cellId) {
+    const expr = this.getExpression(cellId)
+    if (expr) {
+      return expr.syntaxError
+    }
+  }
+
+  recompute(cellId) {
+    // console.log('Recomputing', cellId)
+    const expr = this.getExpression(cellId)
+    // we can only propagate if the expression has been parsed
+    // and the engine has been attached
+    if (expr) {
+      expr.propagate()
+    }
+  }
+
+  _parse(exprStr) {
+    let expr, error
+    if (exprStr) {
+      expr = parseExpression(exprStr)
+      error = expr.syntaxError
+    }
+    return { expr, error }
+  }
+
+  _onEvaluationStarted(expr) {
+    // console.log('evaluation started: ', expr.getSource())
+    super._onEvaluationStarted(expr)
+    const cell = expr._cell
+    if (cell) {
+      this._runtimeErrors[cell.id] = null
+      cell.emit('evaluation:started', expr, cell)
+    }
+  }
+
+  _onEvaluationDeferred(expr) {
+    // console.log('evaluation deferred: ', expr.getSource())
+    super._onEvaluationDeferred(expr)
+    const cell = expr._cell
+    if (cell) {
+      cell.emit('evaluation:awaiting', expr, cell)
+    }
+  }
+
+  _onEvaluationFinished(val, expr) {
+    // console.log('evaluation finished: %s = %s', expr.getSource(), val)
+    super._onEvaluationFinished(val, expr)
+    const cell = expr._cell
+    if (cell) {
+      cell.emit('evaluation:finished', expr, cell)
+    }
+  }
+
+  /*
+    Calling into the context.
+
+    There are different types of calls:
+    - function calls: the arguments are positional (ATM) and passed as array
+    - external cells: arguments are passed as object with
+      names taken from the signature. The context is used to
+      execute the sourceCode, using the arguments object.
+    - chunk: like with external cells, arguments are provided
+      as object. The source code is run in the same way as we know
+      it from notebook cells, such as in Jupyter.
+  */
+  callFunction(funcNode) {
+    const functionName = funcNode.name
+    // ATTENTION: we removed support for 'external cells' (mini + external source code)
+    // as we want to evaluate how far we can get with just
```

Comment:
Worried face :worried:

Reply:
Don't worry too much. If it turns out to be necessary later we bring this back again. 
But this is actually a consequence of your proposal of making external code stronger:better reflection + implicit return :)
Having this makes 'external cells' which are essentially self calling functions pretty obsolete. And having so many concepts for cells/expressions is worrying me personally.

Reply:
Removing this is really worrying me! It's a major departure in terms of the functionality of the entire platform. I'm really surprised.

Reply:
OK, so now I understand this much better after talking to @oliver---- and think it's a great idea to explore - getting rid of the mini special header functions (e.g. `r()`) and instead doing code dependency analysis - looking forward to it!